src/biz/collab/queue.rs

```diff
@@ -0,0 +1,467 @@
+  let meta_map = pending_metas
+    .iter()
+    .map(|meta| (meta.object_id.clone(), meta))
+    .collect::<HashMap<_, _>>();
+
+  // Prepare collaboration data for writing to the database
+  let records = results
+    .into_iter()
+    .filter_map(|(object_id, result)| {
+      if let QueryCollabResult::Success { encode_collab_v1 } = result {
+        meta_map.get(&object_id).map(|meta| PendingWriteData {
+          uid: meta.uid,
+          workspace_id: meta.workspace_id.clone(),
+          object_id: meta.object_id.clone(),
+          collab_type: meta.collab_type.clone(),
+          encode_collab_v1,
+        })
+      } else {
+        None
+      }
+    })
+    .collect::<Vec<_>>();
+
+  // Start a database transaction
+  let mut transaction = collab_cache
+    .pg_pool()
+    .begin()
+    .await
+    .context("Failed to acquire transaction for writing pending collaboration data")
+    .map_err(AppError::from)?;
+
+  // Insert each record into the database within the transaction context
+  for (index, record) in records.into_iter().enumerate() {
+    let params = CollabParams {
+      object_id: record.object_id.clone(),
+      collab_type: record.collab_type,
+      encoded_collab_v1: record.encode_collab_v1,
+    };
+    let savepoint_name = format!("sp_{}", index);
+
+    // using savepoint to rollback the transaction if the insert fails
+    sqlx::query(&format!("SAVEPOINT {}", savepoint_name))
+      .execute(transaction.deref_mut())
+      .await?;
+    if let Err(err) = collab_cache
+      .insert_encode_collab_in_disk(&record.workspace_id, &record.uid, params, &mut transaction)
+      .await
+    {
+      trace!(
+        "Failed to write {} data to disk: {:?}, rollback to save point",
+        record.object_id,
+        err
+      );
+      sqlx::query(&format!("ROLLBACK TO SAVEPOINT {}", savepoint_name))
+        .execute(transaction.deref_mut())
+        .await?;
+    } else {
+      success_write_objects.push(record.object_id);
+    }
+  }
+
+  // Commit the transaction to finalize all writes
+  transaction
+    .commit()
+    .await
+    .context("Failed to commit the transaction for pending collaboration data")
+    .map_err(AppError::from)?;
+  Ok(success_write_objects)
+}
+
+#[inline]
+async fn consume_pending_write_keys(
+  pending_heap: &PendingWriteHeap,
+  maximum_payload_size: usize,
+) -> Vec<String> {
+  const NUM_ITEMS: usize = 5;
+  let mut pending_lock = pending_heap.lock().await;
+  let mut write_items = Vec::with_capacity(NUM_ITEMS);
+  while let Some(item) = pending_lock.pop() {
+    write_items.push(item);
+    if write_items.iter().map(|v| v.data_len).sum::<usize>() > maximum_payload_size {
+      break;
+    }
+    if write_items.len() >= NUM_ITEMS {
+      break;
+    }
+  }
+  drop(pending_lock);
+
+  write_items
+    .iter()
+    .map(|pending| storage_cache_key(&pending.object_id, pending.data_len))
+    .collect()
+}
+
+#[derive(Clone)]
+pub(crate) struct PendingWrite {
```

Comment:
I supposed this is stored in mem. For scalability, I would suggest using redis sorted set

Reply:
done